### PR TITLE
Optimize Bulk Message Parsing and Message Length Parsing

### PR DESCRIPTION
**Pull request by original-brownbear**
Found while profiling the impact of #39286 :
* `findNextMarker` took almost 1ms per invocation during the PMC rally track which comes out 
  * Fixed to be about an order of magnitude faster by using Netty's bulk `ByteBuf` search
* It is unnecessary to instantiate an object (the input stream wrapper) and throw it away, just to read the `int` length from the message bytes
  * Fixed by adding bulk `int` read to BytesReference

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java [hunk 1]
```diff
@@ -276,10 +276,9 @@ public static void readMultiLineFormat(BytesReference data,
     }
 
     private static int findNextMarker(byte marker, int from, BytesReference data, int length) {
-        for (int i = from; i < length; i++) {
-            if (data.get(i) == marker) {
-                return i;
-            }
+        final int res = data.findNextMarker(marker, from, length);
+        if (res > -1) {
```
Comment (s1monw) [inline server/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java]: this should be `!= -1` and assert that it's positive instead?

**Review by s1monw — APPROVED** [review]
LGTM left one comment

**Comment by jpountz** [comment]
Nit: this looks very similar to `String#indexOf` and `List#indexOf`, should we use the same name here? It also looks like we might be able to simplify the signature by dropping `to` and assuming it is always the length of the `BytesReference`?

**Comment by original-brownbear** [comment]
@jpountz you're right, it is always the length for the `to` value here => removed the param + renamed the method to `indexOf` => much nicer I agree :)

File: server/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java (hunk 1 above)
Comment (original-brownbear) [reply server/src/main/java/org/elasticsearch/action/search/MultiSearchRequest.java]: Right, adjusted accordingly :)

**Review by jpountz — CHANGES_REQUESTED** [review]
Let's add some tests to `AbstractBytesReferenceTestCase`? Other than that LGTM.

File: server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java [hunk 2]
```diff
@@ -60,6 +60,29 @@ public static BytesReference bytes(XContentBuilder xContentBuilder) {
      */
     public abstract byte get(int index);
 
+    /**
+     * Returns the integer read from the 4 bytes starting at the given index.
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/common/bytes/BytesReference.java]: maybe clarify that this uses big endian order

**Comment by original-brownbear** [comment]
@jpountz thanks!, adjusted Javadoc and added tests as requested, can you take another look? :)

**Comment by original-brownbear** [comment]
Small issue with the new test my bad, fixing ...

**Comment by original-brownbear** [comment]
@jpountz alright, sorry for the noise above. Now it's green and should be good to review :)

**Review by jpountz — APPROVED** [review]

**Comment by original-brownbear** [comment]
@jpountz thanks!

**Pull request merged**